cajita/unit_test/tstSparseDimPartitioner.hpp

```diff
@@ -0,0 +1,473 @@
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <Cajita_SparseDimPartitioner.hpp>
+#include <Cajita_SparseIndexSpace.hpp>
+#include <Kokkos_Core.hpp>
+
+#include <ctime>
+#include <gtest/gtest.h>
+#include <set>
+
+#include <mpi.h>
+
+using namespace Cajita;
+
+namespace Test
+{
+
+/*!
+  \brief In this test, every cell in the whole domain is registered, so the
+  ground truth partition should be the average partition
+*/
+void uniform_distribution_automatic_rank()
+{
+    // define the domain size
+    constexpr int size_tile_per_dim = 16;
+    constexpr int cell_per_tile_dim = 4;
+    constexpr int size_per_dim = size_tile_per_dim * cell_per_tile_dim;
+    constexpr int total_size = size_per_dim * size_per_dim * size_per_dim;
+
+    // some settings for partitioner
+    float max_workload_coeff = 1.5;
+    int workload_num = total_size;
+    int num_step_rebalance = 100;
+    int max_optimize_iteration = 10;
+    std::array<int, 3> global_cells_per_dim = {
+        size_tile_per_dim * cell_per_tile_dim,
+        size_tile_per_dim * cell_per_tile_dim,
+        size_tile_per_dim * cell_per_tile_dim };
+
+    // partitioner
+    SparseDimPartitioner<TEST_DEVICE, cell_per_tile_dim> partitioner(
+        MPI_COMM_WORLD, max_workload_coeff, workload_num, num_step_rebalance,
+        global_cells_per_dim, max_optimize_iteration );
+
+    // check the value of some pre-computed constants
+    auto cbptd = partitioner.cell_bits_per_tile_dim;
+    EXPECT_EQ( cbptd, 2 );
+
+    auto cnptd = partitioner.cell_num_per_tile_dim;
+    EXPECT_EQ( cnptd, 4 );
+
+    // test ranks per dim
+    auto ranks_per_dim =
+        partitioner.ranksPerDimension( MPI_COMM_WORLD, global_cells_per_dim );
+
+    EXPECT_EQ( ranks_per_dim[0] >= 1, true );
+    EXPECT_EQ( ranks_per_dim[1] >= 1, true );
+    EXPECT_EQ( ranks_per_dim[2] >= 1, true );
+
+    // initialize partitions (averagely divide the whole domain)
+    std::array<std::vector<int>, 3> rec_partitions;
+    for ( int d = 0; d < 3; ++d )
+    {
+        int ele = size_tile_per_dim / ranks_per_dim[d];
+        int part = 0;
+        for ( int i = 0; i < ranks_per_dim[d]; ++i )
+        {
+            rec_partitions[d].push_back( part );
+            part += ele;
+        }
+        rec_partitions[d].push_back( size_tile_per_dim );
+    }
+    partitioner.initializeRecPartition( rec_partitions[0], rec_partitions[1],
```

Comment:
Can the partitioning interface allocate the vectors inside of this array? It would prevent users needing to figure this out with a call to `ranksPerDimension`

Reply:
So the original design is like, the user may probably want to set up the ranks_per_dim manually and initialize the partition in some special way (not just the average way I used in the unit tests). 
But it's true here if the user wants to leave everything to us, this call would be tedious. So I think we can call this `initializeRecPartition` in the constructor to create an average partitioner as initialization. In this way, both `ranksPerDimension`, as well as the `initializeRecPartition`, are hidden from the user. 
How do like it if I do this before merging?

Reply:
Let's get this merged as-is and then do another pull request where we make API updates